File: LeetCode/Monthly POTD/October-2023/343. Integer Break/integerBreak.go
func integerBreak(n int) int {
    dp := make([]int, n+1)
    dp[0] = 0
    dp[1] = 1
    dp[2] = 1

    for i := 3; i <= n; i++ {
        for j := 1; j < i; j++ {
            x := j
            y := i - x
            dp[i] = max(dp[i], max(x*y, dp[y]*x))
        }
    }

    return dp[n]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: LeetCode/Monthly POTD/October-2023/343. Integer Break/integerBreak.cpp
class Solution {
public:
    int integerBreak(int n) {
        vector<int> dp(n + 1, 0);
        dp[0] = 0;
        dp[1] = 1;
        dp[2] = 1;

        for (int i = 3; i <= n; i++) {
            for (int j = 1; j < i; j++) {
                int x = j;
                int y = i - x;
                dp[i] = max(dp[i], max(x * y, dp[y] * x));
            }
        }

        return dp[n];
    }
};